src/libraries/Common/src/System/Text/ValueStringBuilder.cs

```diff
@@ -170,9 +170,10 @@ public void Insert(int index, string? s)
         public void Append(char c)
         {
             int pos = _pos;
-            if ((uint)pos < (uint)_chars.Length)
+            Span<char> chars = _chars;
+            if ((uint)pos < (uint)chars.Length)
             {
-                _chars[pos] = c;
+                chars[pos] = c;
```

Comment:
Both methods are indexing on a Span<char>, correct? Why does this new method avoid a range check?

Reply:
If both the index and span are locals, and you index into those same locals, the JIT avoids emitting an extra range check and helper for the throw.

Reply:
Hmm I might have misunderstood. My original assumption was that the `if ((uint)pos < (uint)chars.Length)` allows the JIT to skip the range check in `chars[pos] = c;`, is that not the case?

Reply:
Right, that is the case after this change.
But that optimization only works on locals. Before, we were reading the field twice (once in the length check and once in the assignment), and by the time we got to the assignment, the JIT couldn't tell that it already checked the length, so it emitted another one.

Reply:
Awesome, that makes sense, thanks for the explanation.

Out of curiosity, do you know what prevents the JIT from optimizing the original case? Is it the potential for the global `_chars` to be accessed by another thread after the if statement? Or is optimizing globals just out of scope for the JIT for some simpler reason?

Reply:
In a general case where you have a field on a struct/class, a different thread could swap out the value from under you between the two reads, so you can't make the assumption that avoiding the second length check is safe.

In this case, this is a field on a ref struct though, so it may be a valid assumption? I'll leave that up to smarter people like @EgorBo though.

Reply:
That's https://github.com/dotnet/runtime/issues/72004